README.md
# Deploy ChatGPT-to-API
[中文文档](./README_zh.md)  English Docs  
These days I found an interesting repo [ChatGPT-to-API](https://github.com/acheong08/ChatGPT-to-API). You can deploy a *fake* API using the web version of ChatGPT.  
However, it's docs are not specific. This repo is to help you easily config and deploy [ChatGPT-to-API](https://github.com/acheong08/ChatGPT-to-API).  

## Features
Original Project Features:  
- Implemented a local Fake API based on a web version of reverse engineering.  
- Supports calling all models supported by the web version (GPT-4 requires Plus subscription).  
- Supports multiple users to prevent triggering rate limits.  
- Implemented APIs:  
  - `/v1/chat/completions`  
  - `/v1/engines`: The original project accessed this api to obtain available models using the official API and key. However, since it relies on the web-based API, it is not very meaningful. Moreover, this API is temporarily unavailable in order to implement custom API keys.  
  -  `/v1/models`: The original project accessed this api to obtain available models using the official API and key. However, since it relies on the web-based API, it is not very meaningful. Moreover, this API is temporarily unavailable in order to implement custom API keys.  

Additional Features I have added:  
- Implemented automatic retrieval of access_token using emails and passwords, and automatically refreshes the access_token upon expiration (only applicable for email password login; Google and Microsoft logins require manual access_token input). The original project only supports single-account one-click login, while this project supports multiple-account one-click access_token retrieval.  
- Implemented authentication. The original project, being based on reverse engineering of the web version, does not require API key verification. However, for added security, I have added a validation feature where you can set the desired API key for request verification. If not needed, you can leave it blank.  
- Supports https. You should only provide certificates and keys and there's no need to use nginx or other reverse proxy.  

## Requirements
- git  
- python3  
- Docker (if deploy in Docker)  
- golang and `go` command in PATH (if deploy in host)  
- One or more ChatGPT accounts  

## Steps
1. Clone this repo to somewhere (suppose `/dcta/`)  
2. Edit the following variables in `run.py`:  
   - `proxy`: format: `host:port`. If you don't need proxy, set it to `""`  
   - `proxy_type`: possible values: `"socks5"` or `"http"`  
   - `accounts`: It's a dictionary of accounts' info. Multiple users are supported. See the example bellow  
   - `server_host`: the host you want ChatGPT-to-API to listen  
   - `server_port`: the port you want ChatGPT-to-API to listen  
   - `custom_API_key`: In theory, since the original project is based on reverse engineering of the web version, there is no need to verify the API key. However, for security purposes, I have added an authentication feature where you can set the desired API key to validate the requests. If you don't need it, you can leave it blank.  
   - `cert_filename`: the filename of the certificate file. Not the path. If you don't need https, set it to `""`  
   - `key_filename`: the filename of the key file. Not the path. If you don't need https, set it to `""`  

![image](https://github.com/Geniucker/Deploy-ChatGPT-to-API/assets/61449208/5c33d3f9-bf21-4a04-af34-579dc6e5fe73)  
3. **If deployed in host**: run `pip3 install -r requirements.txt`  
   **If deployed in docker**: nothing to do in this step  
4. run `build.py` and follow the instructions (You may need to use `sudo` to get the permission of Docker).  
5. run the service:  
   **If deployed in docker**:  
   open a terminal in `/dcta/`. run `docker compose up -d`  
   **If deployed in host**:
   open a terminal in `/dcta/`. run `run.py`  
6. setup https (optional):  
   **If deployed in docker**:  
   modify the `volumes` field in `docker-compose.yml` to replace `./certifications` with the path of your certifications. As the screenshot bellow:  
   ![image](https://github.com/Geniucker/Deploy-ChatGPT-to-API/assets/61449208/2ae9c330-c360-40f1-b741-03c217191e11)  
   **If deployed in host**:  
   Place the certificate and key in or create a symbolic link to the `certifications` folder.  
7. Enjoy~

## FAQ
- Q: What if the access_token of OpenAI expire?  
  A: If this happen, the request to the fake api will cause `401` status code in ChatGPT-to-API. I use python to inspect this code and will regenerate access_token and restart ChatGPT-to-API automatically.  
- Q: Where can I find logs  
  A: This project is divided into two parts. One part is the auxiliary for ChatGPT-to-API, which handles operations like token retrieval and automatic token refresh. Its logs are directly printed on the console.  
  The other part is implemented in Go for authentication and HTTPS. Its logs are located in the `log` folder, specifically in the `authentication.log` file.  
  For deployments within Docker, you can view the container logs to see the logs of the former part. As for the logs of the latter part, you can find them in `log/authentication.log` on the host machine.  

## Contribution
Welcome to raise issues and PRs.  
For bugs, please refer to the log viewing method in the FAQ and provide the logs.  
![log](https://github.com/Geniucker/Deploy-ChatGPT-to-API/assets/61449208/e8472434-780d-4dcc-aaa1-75154c21b917)  


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.1] - 2023-07-04
### Fixed
- failed authentication when custom_API_key is empty (#4) @Geniucker  
- unable to use host proxy when using Docker in Linux (#5) @Geniucker  
- response body is not complete (#7) @Geniucker  


## [2.1.0] - 2023-07-01
### Added
- support for https  


## [2.0.0] - 2020-06-30

### Added
- support for authentication for custom API key  

### Fixed
- the condition of expiration of access_token  


## earlier versions
### Features
- automatically retrieve access_token using emails and passwords, and automatically refreshes the access_token upon expiration (only applicable for email password login  
- easy to deploy  

build.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


from subprocess import getoutput, Popen, PIPE
import os
import sys
import run

def ERROR(msg):
    print("\033[31m[ERROR]\033[0m" + msg)

def testCMD(cmd:str):
    cmd = cmd.split(" ")[0]
    check = False

    if sys.platform.find("win32") != -1:
        if ".exe" != cmd[-4:]:
            cmd += ".exe"
        path = os.environ["PATH"].split(";")
        for i in path:
            if os.path.exists("{}\\{}".format(i, cmd)):
                check = True
                break
    else:
        path = os.environ["PATH"].split(":")
        for i in path:
            if os.path.exists("{}/{}".format(i, cmd)):
                check = True
                break
    return check

# check if run in root of repo
if not os.path.exists("./build.py"):
    ERROR("Please run this script in the root of your cloned repo (Deploy-Chatgpt-to-API).")

# check git command
if not testCMD("git"):
    ERROR("Please install git first.")
    input("Press enter to exit...")
    exit(1)

# clone ChatGPT-to-API
if not os.path.exists("ChatGPT-to-API"):
    choice = ""
    while choice not in ["1", "2"]:
        choice = input(
    """Which method do you want to clone ChatGPT-to-API?
    1. https (you may need to setup proxy for git)
    2. ssh (if you have set up ssh key, ssh is recommended)"""
        )
    if choice == "1":
        os.system("git clone https://github.com/Geniucker/ChatGPT-to-API.git")
    elif choice == "2":
        os.system("git clone git@github.com:Geniucker/ChatGPT-to-API.git")
else:
    print("Try to update ChatGPT-to-API...")
    os.system("cd ChatGPT-to-API && git pull && cd ..")
if not os.path.exists("./ChatGPT-to-API"):
    ERROR("Clone failed. Please check your network and try again.")

choice = ""
while choice not in ["1", "2"]:
    choice = input(
"""Which method do you want to use?
1. Run in docker
2. Run in host
"""
    )

print("Please modify the variable \"proxy\", \"proxy_type\", \"accounts\", \"server_host\", \"server_port\" in run.py")
while "continue" != input("type continue to continue: "):
    pass

print("Your proxy is: {}".format(run.proxy))
print()
print("your accounts are:")
for i in run.accounts:
    print("- {}".format(i))
print()

while "continue" != input("type continue to continue:"):
    pass


# build in docker
if choice == "1":
    if not testCMD("docker"):
        ERROR("Please install docker first.")
        input("Press enter to exit...")
        exit(1)

    # set proxy
    with open("Dockerfile_template", "r") as f:
        dockerfile = f.read()
    if run.proxy != "":
        with open("Dockerfile", "w") as f:
            if run.proxy.split(":")[0] in ["127.0.0.1", "localhost", "::1"]:
                proxy = "host.docker.internal:{}".format(run.proxy.split(":")[1])
            else:
                proxy = run.proxy
            f.write(
                dockerfile.format(
                    "ARG http_proxy={}://{}".format(run.proxy_type, proxy),
                    "ARG https_proxy={}://{}".format(run.proxy_type, proxy)
                )
            )
    else:
        with open("Dockerfile", "w") as f:
            f.write(dockerfile.format("", ""))

    os.system("docker compose down")
    if getoutput("docker image ls").find("chatgpttoapi") != -1:
        os.system("docker image rm chatgpttoapi")
    os.system("docker compose build")

# build in host
elif choice == "2":
    if not testCMD("go"):
        ERROR("Please install go first.")
        input("Press enter to exit...")
        exit(1)

    # set proxy
    if run.proxy != "":
        os.environ["https_proxy"] = run.proxy
        os.environ["http_proxy"] = run.proxy

    os.chdir("ChatGPT-to-API")
    os.system("go build")
    if sys.platform.find("win32") != -1:
        os.system("copy freechatgpt.exe ..")
    else:
        os.system("cp freechatgpt ..")
    os.chdir("../Authenticatior")
    os.system("go build")
    if sys.platform.find("win32") != -1:
        os.system("copy authentication.exe ..")
    else:
        os.system("cp authentication ..")


input("Build finished. Press enter to exit...")


Authenticatior/authentication.go
package main

import (
	"io"
	"log"
	"net/http"
	"os"
)

func authentication(key string, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Printf("Received %s request for %s", r.Method, r.URL.Path)
		if key != "" && r.Method != "OPTIONS" && r.Header.Get("Authorization") != "Bearer "+key {
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			log.Printf("Unauthorized %s request for %s", r.Method, r.URL.Path)
			return
		}
		if r.Header.Get("Authorization") != "" {
			r.Header.Del("Authorization")
		}
		next.ServeHTTP(w, r)
	})
}

func handleRequests(forwardAddr string) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// create new request to forwardAddr
		req, err := http.NewRequest(r.Method, forwardAddr+r.URL.Path, r.Body)
		if err != nil {
			log.Printf("Failed to create request: %s", err)
			http.Error(w, "Bad Gateway", http.StatusBadGateway)
			return
		}
		// copy request headers
		req.Header = r.Header.Clone()
		// send request
		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			log.Printf("Failed to execute request: %s", err)
			http.Error(w, "Bad Gateway", http.StatusBadGateway)
			return
		}
		defer resp.Body.Close()

		// send data back to client
		// copy response headers
		for k, v := range resp.Header {
			w.Header()[k] = v
		}
		// use small buffer to flush immediately
		flusher, _ := w.(http.Flusher)
		buf := make([]byte, 64)
		for {
			n, err := resp.Body.Read(buf)
			if n > 0 {
				w.Write(buf[:n])
				flusher.Flush()
			}
			if err == io.EOF {
				break
			} else if err != nil {
				http.Error(w, "Wrong streaming", http.StatusBadGateway)
				return
			}
		}

		log.Printf("Finished %s request for %s", r.Method, r.URL.Path)
	})
}

func AuthenticationAndForward(listenAddr string, forwardAddr string, key string, certPath string, keyPath string) {
	handler := authentication(key, handleRequests(forwardAddr))

	http.Handle("/", handler)

	log.Printf("Starting HTTP proxy on %s", listenAddr)
	var err error
	if fileExists(certPath) && fileExists(keyPath) {
		err = http.ListenAndServeTLS(listenAddr, certPath, keyPath, nil)
	} else {
		err = http.ListenAndServe(listenAddr, nil)
	}
	if err != nil {
		log.Fatalf("Failed to listen on %s: %s", listenAddr, err)
	}
}

func fileExists(path string) bool {
	if path == "" {
		return false
	}
	_, err := os.Stat(path)
	return !os.IsNotExist(err)
}


run.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from subprocess import getoutput, Popen, PIPE, STDOUT
import os
import sys
import base64
import hashlib
import requests
import shelve
import datetime
from urllib.parse import urlparse, parse_qs
from certifi import where
from datetime import datetime as dt

proxy = "localhost:7890"  # if you don't use proxy, set it to ""
proxy_type = "socks5"  # socks5 or http
accounts = {
    "user1": "password1",
    # "user2": "password2"
}
access_tokens = [
    # "your access token",
    # "your another access token"
]
custom_API_key = ""  # leave it blank if you don't need
server_host = "0.0.0.0"
server_port = "8080"  # if deploy in docker, set it to "8080"
# https (If you don't need, leave it blank)
cert_filename = ""
key_filename = ""


def ERROR(msg):
    print("\033[31m[ERROR]\033[0m" + msg)
def INFO(msg):
    print("\033[32m[INFO]\033[0m" + msg)
def WARNING(msg):
    print("\033[33m[WARNING]\033[0m" + msg)
def generate_code_verifier():
    # 随机生成一个长度为 32 的 code_verifier
    token = os.urandom(32)
    code_verifier = base64.urlsafe_b64encode(token).rstrip(b'=')
    return code_verifier.decode('utf-8')
def generate_code_challenge(code_verifier):
    # 对 code_verifier 进行哈希处理，然后再进行 base64url 编码，生成 code_challenge
    m = hashlib.sha256()
    m.update(code_verifier.encode('utf-8'))
    code_challenge = base64.urlsafe_b64encode(m.digest()).rstrip(b'=')
    return code_challenge.decode('utf-8')
code_verifier = generate_code_verifier()
code_challenge = generate_code_challenge(code_verifier)
class Auth0:
    def __init__(self, email: str = None, password: str = None, access_token: str = None, proxy: str = None, code_verifier: str = None, code_challenge: str = None):
        self.session_token = None
        self.email = email
        self.password = password
        self.session = requests.Session()
        self.code_verifier = code_verifier
        self.code_challenge = code_challenge
        self.expires = None
        self.req_kwargs = {
            'proxies': {
                'http': proxy,
                'https': proxy,
            } if proxy else None,
            'verify': where(),
            'timeout': 100,
        }
        self.access_token = access_token
        self.refresh_token = None
        self.user_agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) ' \
                          'Chrome/109.0.0.0 Safari/537.36'
        try:
            self.load_cache()
        except:
            self.auth()
            self.make_cache()
    def __del__(self):
        if self.refresh_token:
            url = "https://auth0.openai.com/oauth/revoke"
            headers = {
                'User-Agent': self.user_agent
            }
            data = {
                "client_id": "pdlLIX2Y72MIl2rhLhTE9VV9bN905kBh",
                "token": self.refresh_token
            }
            self.session.post(url, headers=headers, data=data, **self.req_kwargs)
    def expire(self):
        self.expires = None
        self.access_token = None
        self.make_cache()
    def make_cache(self):
        if not os.path.exists('cache'):
            os.mkdir('cache')
        with shelve.open('cache/cache', writeback=True) as db:
            db[self.email] = {
                'access_token': self.access_token,
                'expires': self.expires
            }
    def load_cache(self):
        with shelve.open('cache/cache') as db:
            self.access_token = db[self.email]['access_token'],
            self.access_token = self.access_token[0]
            self.expires = db[self.email]['expires']
            INFO('Load cache of {} successfully.'.format(self.email))
    def refresh(self) -> str:
        if self.email is None or self.password is None and self.access_token is not None:
            return self.access_token
        if self.access_token and self.expires and self.expires > dt.now():
            return self.access_token
        elif self.refresh_token:
            url = "https://auth0.openai.com/oauth/token"
            headers = {
                'User-Agent': self.user_agent
            }
            data = {
                "redirect_uri": "com.openai.chat://auth0.openai.com/ios/com.openai.chat/callback",
                "grant_type": "refresh_token",
                "client_id": "pdlLIX2Y72MIl2rhLhTE9VV9bN905kBh",
                "refresh_token": self.refresh_token
            }
            resp = self.session.post(url, headers=headers, data=data, **self.req_kwargs)
            if resp.status_code == 200:
                json = resp.json()
                self.access_token = json['access_token']
                self.expires = dt.utcnow() + datetime.timedelta(seconds=json['expires_in']) - datetime.timedelta(minutes=5)
                self.make_cache()
                INFO('Refresh token of {} successfully.'.format(self.email))
                return self.access_token
            else:
                raise Exception('Error refresh token.')
        else:
            self.auth()
            self.make_cache()
            return self.access_token
    def auth(self) -> str:
        return self.__part_two()
    def __part_two(self) -> str:
        code_challenge = self.code_challenge
        code_verifier = self.code_verifier
        url = 'https://auth0.openai.com/authorize?client_id=pdlLIX2Y72MIl2rhLhTE9VV9bN905kBh&audience=https%3A%2F' \
                '%2Fapi.openai.com%2Fv1&redirect_uri=com.openai.chat%3A%2F%2Fauth0.openai.com%2Fios%2Fcom.openai.chat' \
                '%2Fcallback&scope=openid%20email%20profile%20offline_access%20model.request%20model.read' \
                '%20organization.read%20offline&response_type=code&code_challenge={}' \
                '&code_challenge_method=S256&prompt=login'.format(code_challenge)
        return self.__part_three(code_verifier, url)
    def __part_three(self, code_verifier, url: str) -> str:
        headers = {
            'User-Agent': self.user_agent,
            'Referer': 'https://ios.chat.openai.com/',
        }
        resp = self.session.get(url, headers=headers, allow_redirects=True, **self.req_kwargs)
        if resp.status_code == 200:
            try:
                url_params = parse_qs(urlparse(resp.url).query)
                state = url_params['state'][0]
                return self.__part_four(code_verifier, state)
            except IndexError as exc:
                raise Exception('Rate limit hit.') from exc
        else:
            raise Exception('Error request login url.')
    def __part_four(self, code_verifier: str, state: str) -> str:
        url = 'https://auth0.openai.com/u/login/identifier?state=' + state
        headers = {
            'User-Agent': self.user_agent,
            'Referer': url,
            'Origin': 'https://auth0.openai.com',
        }
        data = {
            'state': state,
            'username': self.email,
            'js-available': 'true',
            'webauthn-available': 'true',
            'is-brave': 'false',
            'webauthn-platform-available': 'false',
            'action': 'default',
        }
        resp = self.session.post(url, headers=headers, data=data, allow_redirects=False, **self.req_kwargs)
        if resp.status_code == 302:
            return self.__part_five(code_verifier, state)
        else:
            raise Exception('Error check email.')
    def __part_five(self, code_verifier: str, state: str) -> str:
        url = 'https://auth0.openai.com/u/login/password?state=' + state
        headers = {
            'User-Agent': self.user_agent,
            'Referer': url,
            'Origin': 'https://auth0.openai.com',
        }
        data = {
            'state': state,
            'username': self.email,
            'password': self.password,
            'action': 'default',
        }
        resp = self.session.post(url, headers=headers, data=data, allow_redirects=False, **self.req_kwargs)
        if resp.status_code == 302:
            location = resp.headers['Location']
            if not location.startswith('/authorize/resume?'):
                raise Exception('Login failed.')
            return self.__part_six(code_verifier, location, url)
        if resp.status_code == 400:
            raise Exception('Wrong email or password.')
        else:
            raise Exception('Error login.')
    def __part_six(self, code_verifier: str, location: str, ref: str) -> str:
        url = 'https://auth0.openai.com' + location
        headers = {
            'User-Agent': self.user_agent,
            'Referer': ref,
        }
        resp = self.session.get(url, headers=headers, allow_redirects=False, **self.req_kwargs)
        if resp.status_code == 302:
            location = resp.headers['Location']
            if location.startswith('/u/mfa-otp-challenge?'):
                if not self.mfa:
                    raise Exception('MFA required.')
                return self.__part_seven(code_verifier, location)
            if not location.startswith('com.openai.chat://auth0.openai.com/ios/com.openai.chat/callback?'):
                raise Exception('Login callback failed.')
            return self.get_access_token(code_verifier, resp.headers['Location'])
        raise Exception('Error login.')
    def __part_seven(self, code_verifier: str, location: str) -> str:
        url = 'https://auth0.openai.com' + location
        data = {
            'state': parse_qs(urlparse(url).query)['state'][0],
            'code': self.mfa,
            'action': 'default',
        }
        headers = {
            'User-Agent': self.user_agent,
            'Referer': url,
            'Origin': 'https://auth0.openai.com',
        }

        resp = self.session.post(url, headers=headers, data=data, allow_redirects=False, **self.req_kwargs)
        if resp.status_code == 302:
            location = resp.headers['Location']
            if not location.startswith('/authorize/resume?'):
                raise Exception('MFA failed.')

            return self.__part_six(code_verifier, location, url)

        if resp.status_code == 400:
            raise Exception('Wrong MFA code.')
        else:
            raise Exception('Error login.')
        
    def get_access_token(self, code_verifier: str, callback_url: str) -> str:
        url_params = parse_qs(urlparse(callback_url).query)

        if 'error' in url_params:
            error = url_params['error'][0]
            error_description = url_params['error_description'][0] if 'error_description' in url_params else ''
            raise Exception('{}: {}'.format(error, error_description))

        if 'code' not in url_params:
            raise Exception('Error get code from callback url.')

        url = 'https://auth0.openai.com/oauth/token'
        headers = {
            'User-Agent': self.user_agent,
        }
        data = {
            'redirect_uri': 'com.openai.chat://auth0.openai.com/ios/com.openai.chat/callback',
            'grant_type': 'authorization_code',
            'client_id': 'pdlLIX2Y72MIl2rhLhTE9VV9bN905kBh',
            'code': url_params['code'][0],
            'code_verifier': code_verifier,
        }
        resp = self.session.post(url, headers=headers, json=data, allow_redirects=False, **self.req_kwargs)

        if resp.status_code == 200:
            json = resp.json()
            if 'access_token' not in json:
                raise Exception('Get access token failed, maybe you need a proxy.')
            self.access_token = json['access_token']
            self.refresh_token = json['refresh_token']
            self.expires = dt.utcnow() + datetime.timedelta(seconds=json['expires_in']) - datetime.timedelta(minutes=5)
            INFO('Get access token of {} successfully.'.format(self.email))
            return self.access_token
        else:
            raise Exception(resp.text)


if __name__=="__main__":
    if proxy.split(":")[0] in ["127.0.0.1", "localhost", "::1"] and os.path.exists("/.dockerenv"):
        proxy = "host.docker.internal:{}".format(proxy.split(":")[1])
    if proxy != "":
        with open("./proxies.txt", "w") as f:
            f.write(proxy)
        proxy = "{}://{}".format(proxy_type, proxy)
    else: proxy = None
    os.environ["SERVER_HOST"] = "0.0.0.0"
    os.environ["SERVER_PORT"] = f"{int(server_port)-1}"
    os.environ["ENABLE_HISTORY"] = "false"
    os.environ["GIN_MODE"] = "release"
    account_objs = []
    for i in accounts:
        account_objs.append(
            Auth0(
                email=i,
                password=accounts[i],
                proxy=proxy,
                code_challenge=code_challenge,
                code_verifier=code_verifier
            )
        )
    for i in access_tokens:
        account_objs.append(
            Auth0(
                access_token=i,
                proxy=proxy,
                code_challenge=code_challenge,
                code_verifier=code_verifier
            )
        )

    # run Authentication
    if not os.path.exists("log"):
        os.mkdir("log")

    cmd = [
        ("./" if sys.platform != "win32" else "") + "authentication" + (".exe" if sys.platform == "win32" else ""),
        f"-listenAddr={server_host}:{server_port}",
        f"-forwardAddr=http://127.0.0.1:{int(server_port)-1}",
        f"-key={custom_API_key}"
    ]
    if cert_filename != "" and key_filename != "" and os.path.exists(f"./certifications/{cert_filename}") and os.path.exists(f"./certifications/{key_filename}"):
        cmd.append(
            f"-certPath=./certifications/{cert_filename}"
        )
        cmd.append(
            f"-keyPath=./certifications/{key_filename}"
        )
    with Popen(cmd, stdout=open("./log/authentication.log", "w") , stderr=STDOUT, shell=False) as authentication:

        # run ChatGPT to API
        while True:
            # config access_tokens
            access_token = []
            for i in account_objs:
                access_token.append(i.refresh())
            for i in range(len(access_token)):
                access_token[i] = access_token[i].strip()
                access_token[i] = "\"{}\"".format(access_token[i])
            with open("./access_tokens.json", "w") as f:
                f.write("[")
                f.write(",".join(access_token))
                f.write("]")


            # run
            cmd = "./freechatgpt"
            try:
                screenData = Popen(cmd, stdout=PIPE, stderr=STDOUT, shell=False)
                while True:
                    line = screenData.stdout.readline().decode("utf-8").strip()
                    if line.find(f"{server_host}:{int(server_port)-1}") != -1:
                        line = line.replace(f"{server_host}:{int(server_port)-1}", f"{server_host}:{server_port}")
                    if line.find("[GIN]") != -1:
                        status = line.split()[5].strip()
                        path = line.split()[12].strip()
                        if status == "200":
                            INFO(line)
                        elif status[0] == "5":
                            WARNING(line)
                            WARNING("detected 5xx, restarting...")
                            screenData.terminate()
                            break
                        elif status == "401" and path == '"/v1/chat/completions"':
                            ERROR(line)
                            ERROR("access_token expired, restarting...")
                            screenData.terminate()
                            for i in account_objs:
                                i.expire()
                            break
                        else:
                            WARNING(line)
                    else:
                        INFO(line)
            except:
                screenData.terminate()
                exit()
        